lxd/storage/drivers/drivers_lvm_utils.go

```diff
@@ -0,0 +1,693 @@
+	}
+
+	return nil, fmt.Errorf("Source is not loop file")
+}
+
+// isLVMNotFoundExitError checks whether the supplied error is an exit error from an LVM command
+// meaning that the object was not found. Returns true if it is (exit status 5) false if not.
+func (d *lvm) isLVMNotFoundExitError(err error) bool {
+	runErr, ok := err.(shared.RunError)
+	if ok {
+		exitError, ok := runErr.Err.(*exec.ExitError)
+		if ok {
+			waitStatus := exitError.Sys().(syscall.WaitStatus)
+			if waitStatus.ExitStatus() == 5 {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// pysicalVolumeExists checks if an LVM Physical Volume exists.
+func (d *lvm) pysicalVolumeExists(pvName string) (bool, error) {
+	_, err := shared.RunCommand("pvs", "--noheadings", "-o", "pv_name", pvName)
+	if err != nil {
+		if d.isLVMNotFoundExitError(err) {
+			return false, nil
+		}
+
+		return false, fmt.Errorf("Error checking for LVM physical volume \"%s\"", pvName)
+	}
+
+	return true, nil
+}
+
+// volumeGroupExists checks if an LVM Volume Group exists and returns any tags on that volume group.
+func (d *lvm) volumeGroupExists(vgName string) (bool, []string, error) {
+	output, err := shared.RunCommand("vgs", "--noheadings", "-o", "vg_tags", vgName)
+	if err != nil {
+		if d.isLVMNotFoundExitError(err) {
+			return false, nil, nil
+		}
+
+		return false, nil, fmt.Errorf("Error checking for LVM volume group \"%s\"", vgName)
+	}
+
+	output = strings.TrimSpace(output)
+	tags := strings.SplitN(output, ",", -1)
+
+	return true, tags, nil
+}
+
+// volumeGroupExtentSize gets the volume group's physical extent size in bytes.
+func (d *lvm) volumeGroupExtentSize(vgName string) (int64, error) {
+	output, err := shared.RunCommand("vgs", "--noheadings", "--nosuffix", "--units", "b", "-o", "vg_extent_size", vgName)
+	if err != nil {
+		if d.isLVMNotFoundExitError(err) {
+			return -1, errLVMNotFound
+		}
+
+		return -1, err
+	}
+
+	output = strings.TrimSpace(output)
+	return strconv.ParseInt(output, 10, 64)
+}
+
+// renameVolumeGroup renames a volume group.
+func (d *lvm) renameVolumeGroup(vgName, newVgName string) error {
+	_, err := shared.TryRunCommand("vgrename", vgName, newVgName)
+	if err != nil {
+		return fmt.Errorf("Error renaming LVM volume group from \"%s\" to \"%s\": %v", vgName, newVgName, err)
+	}
+	d.logger.Debug("Volume group renamed", log.Ctx{"vg_name": vgName, "new_vg_name": newVgName})
+
+	return nil
+}
+
+// countLogicalVolumes gets the count of volumes (both normal and thin) in a volume group.
+func (d *lvm) countLogicalVolumes(vgName string) (int, error) {
+	output, err := shared.RunCommand("vgs", "--noheadings", "-o", "lv_count", vgName)
+	if err != nil {
+		if d.isLVMNotFoundExitError(err) {
+			return -1, errLVMNotFound
+		}
+
+		return -1, err
+	}
+
+	output = strings.TrimSpace(output)
+	return strconv.Atoi(output)
+}
+
+// countThinVolumes gets the count of thin volumes in a thin pool.
+func (d *lvm) countThinVolumes(vgName, poolName string) (int, error) {
```

Comment:
Only has a single caller, any reason to keep it separate?

Reply:
@stgraber the same reason as for `volumeGroupExtentSize`, to avoid distracting the reader of `Delete` (which is already quite complicated logic) with boiler plate parsing of the output of `lvs`.